### PR TITLE
sink(ticdc): translate INSERT automatically when mysql sink disables safe mode

### LINKED ISSUE #3589: MySQL reports Deadlock when TiCDC replicating data
### What did you do?

In 1 TiDB, 1 TiKV, 1 TiCDC cluster, set up a changefeed that replicates all tables to downstream MySQL.

MySQL may reports deadlock error.

Reproduce:

* Extends TiDB GC lifetime to 2400h.
* Load data to TiDB, tiup bench tpcc prepare and tiup bench tpcc run
* Create a changefeed `cfa` whose start-ts = TiDB gc safepoint, and replicates data to MySQL.
* Wait checkpoint lag close to 10s.
* Remove changefeed `cfa` and create a changefeed `cfb` use the same config as `cfa`.

We can find deadlock errors in TiCDC log.

MySQL deadlock log: [deadlock.log](https://github.com/pingcap/ticdc/files/7586652/deadlock.log)

Deadlock table schema

```sql
CREATE TABLE `order_line` (
  `ol_o_id` int(11) NOT NULL,
  `ol_d_id` int(11) NOT NULL,
  `ol_w_id` int(11) NOT NULL,
  `ol_number` int(11) NOT NULL,
  `ol_i_id` int(11) NOT NULL,
  `ol_supply_w_id` int(11) DEFAULT NULL,
  `ol_delivery_d` datetime DEFAULT NULL,
  `ol_quantity` int(11) DEFAULT NULL,
  `ol_amount` decimal(6,2) DEFAULT NULL,
  `ol_dist_info` char(24) DEFAULT NULL,
  PRIMARY KEY (`ol_w_id`,`ol_d_id`,`ol_o_id`,`ol_number`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

### What did you expect to see?

No deadlock error.

### What did you see instead?

MySQL Deadlock error.

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
v5.2.1
```

TiCDC version (execute `cdc version`):

```console
v5.2.1
```

MySQL version

```
5.7.36
```

### LINKED ISSUE #5611: An enhancement of MySQL sink to turn off safe-mode automatically
### Is your feature request related to a problem?

By default, TiCDC translates upstream `INSERT/UPDATE` into `REPLACE` (the safe-mode) which is slow and may cause deadlock in MySQL.

  

### Describe the feature you'd like

When TiCDC replicates a table, it can turn off safe-mode automatically when there is no duplicated events that has been replicated before.

TiCDC may output duplicated events when a table is scheduled to another node. The table have to replicate start from the global checkpoint (a checkpoint of the slowest tables) which it may has already replicated.

So turn off safe-mode, it has to know a ts that it has **not** replicated before. The trick is simple, it takes the latest ts from PD after it starts replicating, so any txn that is greater than the tso, it must have never replicated before. after the table see such txn, it can safely turn off safe-mode.

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: close #5611 close #3589

### What is changed and how it works?

See details in #5611.

#### Tests 

TBD

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Tests shows this PR doubles MySQL sink throughputs when its downstream is TiDB.

| master vs this PR |
| - |
| <center>**Downstream TiDB**</center></br> image |

##### Will it cause performance regression or break compatibility?

No

##### Do you need to update user documentation, design documentation or monitoring documentation?

No

### Release note 

Not ready yet.

```release-note
None
```

### CODE REVIEW

**Review by zhaoxinyu — APPROVED** [review]

File: cdc/sink/mysql/mysql_test.go [hunk 1]
```diff
@@ -1499,6 +1498,74 @@ func TestExecDMLRollbackErrRetryable(t *testing.T) {
 	require.Nil(t, err)
 }
 
+func TestMysqlSinkNotRetryErrDupEntry(t *testing.T) {
+	errDup := mysql.NewErr(mysql.ErrDupEntry)
```
Comment (hi-rustin) [inline cdc/sink/mysql/mysql_test.go:1502]:
```suggestion
    t.Parallel()

	errDup := mysql.NewErr(mysql.ErrDupEntry)
```

File: cdc/sink/mysql/mysql.go [hunk 2]
```diff
@@ -697,6 +695,14 @@ func (s *mysqlSink) prepareDMLs(rows []*model.RowChangedEvent, bucket int) *prep
 	rowCount := 0
 	// translateToInsert control the update and insert behavior
 	translateToInsert := s.params.enableOldValue && !s.params.safeMode
+	for _, row := range rows {
+		if !translateToInsert {
+			break
+		}
+		// It can be translated in to INSERT, if the row is committed after
+		// we start replicating the table.
```
Comment (hi-rustin) [inline cdc/sink/mysql/mysql.go:703]: Could you please add comment to explain why?

File: cdc/sink/mysql/mysql_test.go (hunk 1 above)
Comment (hi-rustin) [inline cdc/sink/mysql/mysql_test.go:1502]:
```suggestion
    t.Parallel()

	errDup := mysql.NewErr(mysql.ErrDupEntry)
```

Comment (overvenus) [reply cdc/sink/mysql/mysql_test.go:1502]: Can not parallel, it changes `GetDBConnImpl`, and will affect others tests.

Comment (hi-rustin) [reply cdc/sink/mysql/mysql_test.go:1502]: Got it!

**Review by hi-rustin — APPROVED** [review]

**Pull request merged**